### ISSUE TITLE
bCNC - Sender.py coding help

### ISSUE THREAD

**Issue by toktuff** [vlachoudis/bCNC#971]
Hi

This is a fairly technical "issue".

I would love to get some feedback from a more seasoned bCNC developer. (@vlachoudis)

Currently I'm toying around with a basic implementation of jogging using a gamepad. I've got it working but not very reliable. The use of the new JogCancel seems a bit subtle (https://github.com/gnea/grbl/issues/95). To recap that issue, the best way to reliably know that a jog cancel has been processed is to: jog jog jog, jog cancel, G4P0 => wait for "OK".

Question 1:
How do I know that a given "OK" relates to a specific command. To me it appears that the serial.write is mostly a fire and forget operation.

Question 2:
Now, given that Sender.py is running on a separate thread and they Jogcancel will be initiated from the main thread. How should I get notified of this? I'm considering the option of passing a tuple with an event object. I.e. ("g4p0", my_event) and let the Sender do my_event.set()

Perhaps I could simply wait until it's idle (i.e. sendGcode((WAIT,)).

Any ideas?

**Comment by zeevy** [comment]
I too have this jog cancellation issue. my work around is when you are sending the continuous jog commands to grbl then never fill the planner buffer completely. 
Jog cancellation may not work if you have jog commands left in RX buffer.

I do like
```
if(PlannerBuffer > 5){
 sendJogCommand();
}else{
 ignoreTheJogCommand();
}
```

This still fails in some edge cases since I use grbl status report for getting the planner buffer size. (should calculate and track this on GUI side)

**Comment by hdo** [comment]
Sad to see that bCNC hasn't implemented the 'new' jogging feature of grbl 1.1 yet :-(

**Comment by toktuff** [comment]
Yes, the new jogging feature has some benefits.

I was thinking something similar. There is also a possible race condition in the code. If you dig into Sender.py it can happen that tosend contains a jog command on the serial send thread and if you send a jogcancel on the main thread, this isn't really easy to handle.

The way I got it to work semi-robustly was to clear the queue in bCNC, send jogcancel to grbl, send g4p0 and wait for idle (using a condition variable).

I still managed to hit the race condition described above on occasion.

**Comment by hdo** [comment]
@toktuff 

Yeah i saw that in the code also. I was think about something similar to your solution. I'm thinking about handling jog commands in a separate thread or queue. Problem here is that bCNC has its on queue which need to be cleared first what you also did.

**Comment by toktuff** [comment]
One idea I've been thinking of is to add a second queue to Sender.py. I.e. something like real_time_queue which has priority over the regular queue. This way all serial communication would happen in a single thread and can be scheduled properly. I guess one drawback it will introduce lag in the real time commands. Not sure if it would be significant.
